src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+     * @returns true if successful.
+     */
+    public detach(): boolean {
+        if (!super.detach()) {
+            return false;
+        }
+
+        Object.keys(this._controllers).forEach((controllerId) => {
+            this._detachController(controllerId);
+        });
+
+        return true;
+    }
+
+    /**
+     * Will get the mesh under a specific pointer.
+     * `scene.meshUnderPointer` will only return one mesh - either left or right.
+     * @param controllerId the controllerId to check
+     * @returns The mesh under pointer or null if no mesh is under the pointer
+     */
+    public getMeshUnderPointer(controllerId: string): Nullable<AbstractMesh> {
+        if (this._controllers[controllerId]) {
+            return this._controllers[controllerId].meshUnderPointer;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Get the xr controller that correlates to the pointer id in the pointer event
+     *
+     * @param id the pointer id to search for
+     * @returns the controller that correlates to this id or null if not found
+     */
+    public getXRControllerByPointerId(id: number): Nullable<WebXRInputSource> {
+        const keys = Object.keys(this._controllers);
+
+        for (let i = 0; i < keys.length; ++i) {
+            if (this._controllers[keys[i]].id === id) {
+                return this._controllers[keys[i]].xrController || null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * This function sets webXRControllerPointer Selection feature that will be disabled when
+     * the hover range is reached for a mesh and will be reattached when not in hover range.
+     * This is used to remove the selection rays when moving.
+     * @param farInteractionFeature the feature to disable when finger is in hover range for a mesh
+     */
+    public setfarInteractionFeature(farInteractionFeature: Nullable<IWebXRFeature>) {
+        this._farInteractionFeature = farInteractionFeature;
+    }
+
+    /**
+     * Filter used for near interaction pick and hover
+     */
+    private nearPickPredicate(mesh: AbstractMesh): boolean {
+        return mesh.isEnabled() && mesh.isVisible && mesh.isPickable && mesh.isNearPickable;
+    }
+
+    /**
+     * Filter used for near interaction garb
+     */
+    private nearGrabPredicate(mesh: AbstractMesh): boolean {
+        return mesh.isEnabled() && mesh.isVisible && mesh.isPickable && mesh.isNearGrabbable;
+    }
+
+    private readonly _hoverRadius = 0.1;
+    private readonly _pickRadius = 0.03;
+    private _indexTipQuaternion = new Quaternion();
+    private _indexTipOrientationVector = Vector3.Zero();
+
+    protected _onXRFrame(_xrFrame: XRFrame) {
+        Object.keys(this._controllers).forEach((id) => {
+            // only do this for the selected pointer
+            const controllerData = this._controllers[id];
+            if (!this._options.enablePointerSelectionOnAllControllers && id !== this._attachedController) {

```

Comment:
>enablePointerSelectionOnAllControllers [](http://example.com/codeflow?start=31&length=38)

nit: some naming doesn't make as much sense in the strictly near interaction context, as it talks about the pointer. Maybe something more in-line with the new class name like enableNearInteractionOnAllControllers?